STYLE.rst

```diff
@@ -0,0 +1,515 @@
+      MyClass(MyClass const &) = default;
+      // Move constructor.
+      MyClass(MyClass &&) = default;
+      // Copy assignment operator.
+      MyClass & operator=(MyClass const &) = default;
+      // Move assignment operator.
+      MyClass & operator=(MyClass &&) = default;
+      // Destructor.
+      ~MyClass() = default;
+  }; /* end class MyClass */
+
+C++ Encapsulation
+====
+
+Prefer encapsulated classes over POD struct so that we always provide
+accessors. We provide accessors for even scalars of fundamental types.
+
+.. code-block:: cpp
+
+  class MyPowerHouse
+  {
+
+  public:
+
+      void calculate_internal_data();
+
+      // Use the same-name style for accessors.
+      double internal_value() const { return m_internal_value; }
+      double & internal_value() { return m_internal_value; }
+
+      // It may be good to have a blank line between accessor pairs.
+      SimpleArray<double> const & internal_data() const { return m_internal_data; }
+      SimpleArray<double> & internal_data() { return m_internal_data; }
+
+  private:
+
+      double m_internal_value = 0.0;
+      SimpleArray<double> m_internal_data;
+
+  }; /* end class MyPowerHouse */
+
+Prefer Same-Name Accessors
+----
+
+(Python does not need accessors. Do not add accessors in Python code.)
+
+Prefer same-name accessors because we expose a lot of internal containers:
+
+.. code-block:: cpp
+
+  // Getter is const and return a copy of a fundamental type.
+  double internal_value() const { return m_internal_value; }
+  // Setter is non-const and return a reference.
+  double & internal_value() { return m_internal_value; }
+
+  // Getter is const and return a const reference of a non-fundamental type.
+  SimpleArray<double> const & internal_data() const { return m_internal_data; }
+  // Setter is non-const and return a reference.
+  SimpleArray<double> & internal_data() { return m_internal_data; }
+
+Sometimes we may use the getter-and-setter style to supplement the same-name
+accessors:
+
+.. code-block:: cpp
+
+  // Getter is const and return a copy of a fundamental type.
+  double get_internal_value() const { return m_internal_value; }
+  // Setter takes
+  void set_internal_value(double v) { m_internal_value = v; }
+
+  // Getter is const and return a const reference of a non-fundamental type.
+  SimpleArray<double> const & internal_data() const { return m_internal_data; }
+  // Setter is non-const and return a reference.
+  SimpleArray<double> & internal_data() { return m_internal_data; }
+
+It is OK for accessors of the same-name and getter-and-setter styles to be
+available for the same member datum, but we should only do it when necessary.
+
+C++ Ending Mark
+====
+
+Add ending marks to classes and namespaces.  They are usually too long (across
+hundreds of lines) to keep track of.
+
+.. code-block:: cpp
+
+  namespace modmesh
+  {
+
+  class MyClass
+  {
+      // Code.
+  }; /* end class MyClass */
+
+  } /* end namespace modmesh */
+
+Copyright Notice
```

Comment:
I think copyright could be just a single line refering to the copyright file. It's annoying that a file has tens of line of code just for the copyright.

Reply:
I need the copyright texts to be in all files.

Reply:
I wonder what is the difference? According to what I read from the clean code book, we shouldn't copy the full copyright text everywhere.

Reply:
I need the texts in all files but would not like to provide the reasons.  I cannot change the requirement now but when I can explain more when I can.

The code has different use cases.  Although the code is open source, many of the use cases may not be open.

Reply:
I have difficulty understanding why the amount of comments or notice left in a file may have anything to do with productivity, and do not understand why "clean code" is a support for having short copyright notice.  But I am interested in your argument.

Reply:
It matters the productivity because skipping the copyright while reading code takes time and copying the copyright also takes time. It also distracts people's attention from the code-relevant comments.

Reply:
I disagree that copyright notice distracts a developer, especially after we have written down the style guide.  It is fairly easy to get used to skipping 25-30 lines of code in the beginning of the files after reading around 5 files of that kind.

A developer only gets distracted if they do not plan to spend time reading some files in the code base.

Reply:
I realized my mistake. "Clean Code" advises against using multi-line comments at the beginning of a file for some cases but doesn't mention anything about copyright comments. The book "CODE 2" does discuss copyright comments, providing an example of a one-line copyright notice, but it doesn't state that long copyright comments are undesirable.

Reply:
It makes sense for a book to use a short form of copyright notice because paper-printing costs.  If one line suffices for demonstration it does not need to cover the full license content.  Moreover, a full content needs to pick a specific license agreement and the decision clearly goes outside the scope of a book for software engineering.

While we will not discuss the reasons behind the requirement of add the full license content in the file header, we can still discuss the pros and cons.  The discussion and analysis benefit more than taking a context-free conclusion from a book.

Reply:
We must be careful on taking conclusions from a reference.  Even Newton's laws of motion has a constraint on applicability.  It is usually better to take an argument rather than (just) a conclusion out of a reference.  It is not uncommon for a living project to contain richer context than a book.

Scientific and high-performance computing carries much more context than the software engineering, which does not take [engineering](https://www.abet.org/about-abet/) into account.  This fascinates for it takes more analytical faculty.

Reply:
I see. Good to know :)